confluent_kafka/avro/serializer/__init__.py

```diff
@@ -15,11 +15,39 @@
 # limitations under the License.
 #
 
+import avro
+import avro.io
+import io
+import logging
+import struct
+import sys
+import traceback
+
+from confluent_kafka.avro.schema import GenericAvroRecord, get_schema
+from confluent_kafka.avro.error import ClientError
+
+log = logging.getLogger(__name__)
+
+MAGIC_BYTE = 0
+
+HAS_FAST = False
+try:
+    from fastavro import schemaless_reader, schemaless_writer
+
+    HAS_FAST = True
+except ImportError:
+    pass
+
 
 class SerializerError(Exception):
     """Generic error from serializer package"""
 
-    def __init__(self, message):
+    def __new__(cls, message, is_key=False):
```

Comment:
I don't think you'd ever want to have different handlers for key vs value serialization errors, so I don't think you need two types - just `SerializerError` would be better I think, with a property to indicate whether it's a key or not.

Also, you've used `__new__` again here. This seems like a bad pattern to me - the benefit of doing it your way is it makes usage terser, but in general, this is at the expense of giving up on the idea of knowing WTF you're going to get on instantiation without looking at the implementation. I'm not familiar with working in languages that have this capability, but can't see how this tradeoff is a good one.

Reply:
Bear with me here as I know this goes against all my other arguments where I say I don't want to let Avro dictate python's serde API... Where it makes sense I'm trying to make the new serde api match the behavior of the existing Avro Producer and Consumer. For this reason we differentiate between a key and value serializer. I personally do not have strong feelings for either approach. If you on the other hand feel strongly that this should just be one exception type we can back it out. 

As for `__new__` I honestly don't feel this is too terrible when the resulting object is still the same class. There are no real surprises here, I wanted a Serialization exception I got one. This one is just a tad more specific than the other but they will both be caught by except SerializerError. If we decide to just return one SerializationError type however there is no need for this at all. Anyway let me know what you think, as I said I don't have particularly strong feelings either way. Now that you know my intent I'll defer to your better judgement here.